host/ProtocolGateway.Host.Fabric.FrontEnd/configuration/ServiceFabricConfigurationProvider.cs

```diff
@@ -0,0 +1,112 @@
+        /// </summary>
+        readonly ImmutableDictionary<string, string> configurationValues;
+        #endregion
+        #region Constructors
+        /// <summary>
+        /// The settings provider used to bridge Service Fabric configuration information and DotNetty
+        /// </summary>
+        /// <param name="traceId">The unique identifier used to correlate debugging and diagnostics messages</param>
+        /// <param name="componentName">The component name used for debug and diagnostics messages</param>
+        /// <param name="logger">The service fabric logger to be used when writing debug and diagnostics information</param>
+        /// <param name="gatewayConfiguration">The gateway configuration used to get current configuration information for DotNetty</param>
+        /// <param name="iotHubConfiguration">The IotHub Client configuration used to get current configuration information for DotNetty</param>
+        /// <param name="mqttConfiguration">The MQTT configuration used to get current configuration information for DotNetty</param>
+        public ServiceFabricConfigurationProvider(Guid traceId, string componentName, IServiceLogger logger, GatewayConfiguration gatewayConfiguration, IoTHubConfiguration iotHubConfiguration, MqttServiceConfiguration mqttConfiguration)
+        {
+            this.componentName = componentName;
+            this.logger = logger;
+
+            this.logger.Verbose(traceId, this.componentName, "Initializing configuration provider.");
+
+            var baseProperties = new Dictionary<string, string>();
+
+            foreach (PropertyInfo element in typeof(GatewayConfiguration).GetProperties(BindingFlags.Public | BindingFlags.Instance))
+            {
+                if (!element.PropertyType.IsArray)
+                {
+                    object value = element.GetValue(gatewayConfiguration, null);
+                    baseProperties.Add(element.Name, element.PropertyType == typeof(string) ? value as string : value.ToString());
+                }
+            }
+
+            foreach (PropertyInfo element in typeof(MqttServiceConfiguration).GetProperties(BindingFlags.Public | BindingFlags.Instance))
+            {
+                if (!element.PropertyType.IsArray)
+                {
+                    object value = element.GetValue(mqttConfiguration, null);
+                    baseProperties.Add(element.Name, element.PropertyType == typeof(string) ? value as string : value.ToString());
+                }
+            }
+
+            foreach (PropertyInfo element in typeof(IoTHubConfiguration).GetProperties(BindingFlags.Public | BindingFlags.Instance))
+            {
+                if (!element.PropertyType.IsArray)
+                {
+                    object value = element.GetValue(iotHubConfiguration, null);
+                    baseProperties.Add($"IotHubClient.{element.Name}", element.PropertyType == typeof(string) ? value as string : value.ToString());
```

Comment:
I was torn with this myself. I was debating on using the type descriptor etc. but thought it may create more confusion etc. with no real value. If you have a preferred direction here I am happy to change it